### DISCUSSION TITLE
Does this project implement also the OTA specificatiom of matter

### DISCUSSION THREAD

**Discussion by dberardo** [mfucci/node-matter#178] [Ideas]
I am still new to Matter, but I saw that device library spec defined an OTA client/server procedure.
Is it possible to use this nodes project as a Middleware for that procedure?

If so, can this nodes application achieve OTA on client/server alone, or does it need a previous installation of any other OTA client/server framework/agents (like Mender)?

Thanks for providing clarity.

**Comment by mfucci** [comment]
Hi dberardo,

I haven't had the chance to look on how Matter is handling OTA, but my best guess is that: yes, node-matter could be used as a Middleware for OTA.
From the specs (p 731, section 11.19), node-matter can advertise itself as a OTA Provider" and other Matter devices will automatically connect to it to check if a new firmware is available.

There is no need for a previous installation of any other OTA client/server framework/agents, since OTA is part of the Matter protocol.

Howver, OTA is not implemented yet in node-matter
The logic seems fairly simple, it is a new cluster with 2 commands:

1. QueryImage
2. ApplyUpdate

I don't know when I will get around to implement it but I can give you the pointers on how to implement it, if you are up for the challenge.
The steps will roughly be:

1. add the definition for the new cluster
2. implement a dummy QueryImage command, just logging what the OTA Requestor is requesting
3. setup an environment where one device is requesting an OTA update
4. implement the real QueryImage / ApplyUpdate commands

Thanks!
- Marco

**Reply by dberardo-com** [reply]
> There is no need for a previous installation of any other OTA client/server framework/agents, since OTA is part of the Matter protocol.

this means that in order for node-matter to add OTA as a Matter-compliant feature, then it needs to implement the actual "update/rollback" functionality itself, without relying on 3rd party tools ? Is this even possible, knowing that embedded systems and such devices will usually require Firmware updates ? can those updates be carried out via NodeJS ? seems to me too much of a "low-level" operation for nodejs, doesnt it ?

Or are you talking just about the OTA-Server (Provider) part of the spec here? In this case i assume you mean using node-matter as a sort of replacement for systems like Hawkbit https://github.com/eclipse/hawkbit ? Or are the features offered by Hawkbit way beyond what is required by the Matter OTA Provider spec?

**Comment by mfucci** [comment]
Hi dberardo,

Not sure I follow.
Can you describe your use case in more details?

For OTA updates, there are two distinct independent flows:

1. node-matter code can be updated using OTA
2. node-matter is used to upgrade other Matter in your home over OTA

From your original question, I thought you were interested flow 2: updating other devices, while I have the impression you are now mentioning flow 1: updating node-matter version.

Flow 2 should be pretty straight forward.

Fro flow 1: node-matter is meant to be a library, so not sure if it makes sense to update it over OTA since it will probably be modified for your use case. It will be much easier to update it through npm or git directly. That said, we will implement the OTA flow at some point so if someone want to release official firmwares for their creation, they can be updated using OTA.
I don't see any reason why node.js cannot do that: the old process launches a new process running the new code and closes itself.

Thanks!
- Marco

**Reply by dberardo-com** [reply]
i see what you mean now, you mean that the system to be upgraded is not the actual machine (node) on which node-matter is running, but the actual node-matter code alone. 

In general my use case is that i have one sort of "master node" which schedules updates (like a Hawkbit or Mender server) and a client / agent system on the field devices that verifies the update and performs it. In case of failure, the client system is rolled back to the previous version.

Now, when i talk about field device upgrades, most of the times i am not referring to just a single application running on those systems, but on the whole firmware. That's why the OTA agents running on the clients usually need to interact with the boot loader of the kernel, so that they can download the new Image version and make the system boot from that new image. This is what agents like SWUpdate and Mender (client) essentially do. 

So here i am talking about very different system update procedures and not just, say, "upgrade python from 3.9 to 3.10" or so, but more like "Change the sudoers file, the config file of application XYZ, upgrade the USB driver, upgrade linux packages A,B,C, etc."

I am not sure if Matter OTA procedure is meant to go that deep, but this sort of upgrade is what i usually expect in the context of IoT environments.

**Reply by Apollon77** [reply]
I think "OTA" is very specific to the "device" implementation. I think the usual "OTA" refers to "the firmware of the device is being updated": This would contain the" device logic parts"  but mostly this is hard bound to the "whole device image" including Operating system and such.

If you want ti use node-matter as a "OTA Provider" then in fact node-matter needs to deliver whatever "OTA package" that the device expects and needs. So if you have an ESP32 with C-code based matter implementation this most likely is binary stuff including "runtime" stuff. If we talk about "node-matter based Node.js matter client devices" then it could (!!) be updating just the JS code, but in fact this is never realistic because you also have node_modules and other additional content (not start to talk about nodejs binary itself and such), so I would also see some whatever OTA package (and a mechcnism to apply it and reboot the device, so it will always be device specific.

Thinking it lie taht brings me to the point that node-matter could/should provide a way to get the "OTA new image to be applied" - whatever this is in detail,  would see it as a binary blob, and implement the sending mechanisms. ALl the rest is very devcie specific

Or do I oversee something?

**Reply by dberardo-com** [reply]
i cannot answer your question in a "protocol-compliant" manner, because i havent read carefully that part of the specification myself ... but i would expect Matter to provide clarification on the extent of the OTA procedure on the client devices ...

but lets think about maybe one of the most simple update commands: updating date and time on the device (or maybe changing the timezone / NTP server). Would the node-matter OTA procedure described by you be able to deliver this use case ?

**Reply by Apollon77** [reply]
In fact after having a fly over the Specs they talk about "Software Images"... so I think OTA is really meaned as "Firmware update" and not "update a config". That means that each config update indeed is a complete new image, also because the comparism is done by a "version number"

**Reply by dberardo-com** [reply]
thanks for the clarification.

Is this sort of implementation something you are willing to undertake within the node-matter project? And is so, will it only involve the OTA Provider implementation or also the OTA Requester one ?

**Reply by Apollon77** [reply]
> you are willing to undertake within the node-matter project?

As mfucci stated above this indeed could be done somewhen, but my honest opinion is that this might be a topic once all the basics work and is stable... then it depends on priorities and what makes sense and is wished by the majority of the community.

**Reply by mfucci** [reply]
To clarify here:

- "master node" which schedules updates: node-matter should be able to act as one when the OTA-provider cluster is implemented
- "client / agent system on the field": the OTA Matter update will upload a new "firmware" to the device. What this "firmware" contains is up to the manufacturer. If this client is running node-matter, it can contain a new version of node-matter. But it can also contain a list of .deb files and a script to update the full OS and reboot the machine. node.js can run system commands so can upgrade / reboot the whole OS if one process is running as root. But if your client on the field has network connectivity, you can simply scheduled an automated maintenance so you don't have to worry about updating the OS through Matter. This is all out-of-scope for node-matter...

**Comment by dberardo-com** [comment]
A related question: what is the different scope/aim of the node-matter project compared to this other official repo https://github.com/project-chip/matternode ?

**Reply by Apollon77** [reply]
see https://github.com/mfucci/node-matter/discussions/134
